src/contrib/mlir/lowerer.cpp

```diff
@@ -530,6 +530,102 @@ namespace
         return matchSuccess();
     }
 
+    REWRITER(NGConcatOp)
+    {
+        auto concat = cast<NGConcatOp>(op);
+        auto loc = concat.getLoc();
+
+        ScopedContext scope(rewriter, loc);
+
+        // Create Value for result, and extract type info.
+        Value* result = m_pass.buildOutputDefs(op, rewriter)[0];
+        NGRAPH_CHECK(result, "Unexpected null result in ConcatOp");
+
+        auto result_ty = result->getType().dyn_cast<MemRefType>();
+        NGRAPH_CHECK(result_ty, "Unexpected non-memref result type");
+
+        Type elem_ty = result_ty.getElementType();
+
+        // Retrieve concatenation axis.
+        auto concatenation_axis_attr = op->getAttrOfType<IntegerAttr>("concatenation_axis");
+        NGRAPH_CHECK(concatenation_axis_attr, "Missing concatenation_axis in ConcatOp");
+
+        int64_t concatenation_axis = concatenation_axis_attr.getInt();
+
+        // Create views/handles to write into result.
+        MemRefView v_res(result);
+        auto rank = v_res.rank();
+        IndexedValue iv_res(result);
+
+        // For each operand, generate a separate loop to copy into the target slice of "result".
+        // We'll keep track of the slice offsets via concatenation_axis_pos.
+        IndexHandle concatenation_axis_pos(index_t(0));
+
+        for (auto& operand : operands)
+        {
+            // Retrieve and check some type info.
+            NGRAPH_CHECK(operand, "Unexpected null operand in ConcatOp");
+            auto operand_ty = operand->getType().dyn_cast<MemRefType>();
+            NGRAPH_CHECK(operand_ty, "Unexpected non-memref operand type");
+            NGRAPH_CHECK(elem_ty == operand_ty.getElementType(), "Types mismatch in ConcatOp");
+
+            // TODO(amprocte): Check consistency of operand shape (rank, dims other than at
```

Comment:
That can be achieved via a call to op->verify(). I think it is safe to assume the code is already verified by calling Module->verify() beforehand.

Reply:
I think some of the `NGRAPH_CHECK`s above could probably go away as well, right?

Reply:
yes.

Reply:
I've removed `NGRAPH_CHECK`s that I _think_ are unnecessary given the assumption that the module has been verified. You might take a second look to see if I'm overdoing it.

Reply:
We don't have to get crazy about checks but some checks would be needed. Module verification happens all at once before lowering and it might be disabled in some compilation modes in the future (same as in LLVM). We would also need to check for nullptrs (KW will complain) and the own lowerer may leave ops and types in an inconsistent way. I hit similar memref type asserts before. In general, I would say that checking for nullptrs and expected invariants, such as types, should be fine

Reply:
@dcaballe Makes sense, I'll put some checks in for nullptrs and type assumptions.